buildSrc/src/main/groovy/net/minecraftforge/forge/tasks/checks/CheckPatches.groovy

```diff
@@ -0,0 +1,177 @@
+package net.minecraftforge.forge.tasks.checks
+
+import groovy.transform.CompileStatic
+import org.gradle.api.file.DirectoryProperty
+import org.gradle.api.provider.ListProperty
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.InputDirectory
+import org.gradle.api.tasks.Optional
+
+import java.nio.file.Files
+import java.nio.file.Path
+import java.util.regex.Pattern
+
+@CompileStatic
+abstract class CheckPatches extends CheckTask {
+    @InputDirectory abstract DirectoryProperty getPatchDir()
+    @Input @Optional abstract ListProperty<String> getPatchesWithS2SArtifact()
+
+    @Override
+    void check(Reporter reporter, boolean fix) {
+        final patchDir = getPatchDir().get().asFile.toPath()
+        Files.walk(patchDir).withCloseable {
+            it.filter(Files.&isRegularFile).forEach { path ->
+                final String relativeName = patchDir.relativize(path).toString()
+                verifyPatch(path, reporter, fix, relativeName, patchesWithS2SArtifact.get().contains(relativeName.replace('\\', '/')))
+            }
+        }
+    }
+
+    void verifyPatch(Path patch, Reporter reporter, boolean fix, String patchPath, boolean hasS2SArtifact) {
+        final hunk_start_pattern = Pattern.compile('^@@ -[0-9,]* \\+[0-9,_]* @@$')
+        final white_space_pattern = Pattern.compile('^[+\\-]\\s*$')
+        final import_pattern = Pattern.compile('^[+\\-]\\s*import.*')
+        final field_pattern = Pattern.compile('^[+\\-][\\s]*((public|protected|private)[\\s]*)?(static[\\s]*)?(final)?([^=;]*)(=.*)?;\\s*$')
+        final method_pattern = Pattern.compile('^[+\\-][\\s]*((public|protected|private)[\\s]*)?(static[\\s]*)?(final)?([^(]*)[(]([^)]*)?[)]\\s*[{]\\s*$')
+        final class_pattern = Pattern.compile('^[+\\-][\\s]*((public|protected|private)[\\s]*)?(static[\\s]*)?(final[\\s]*)?(class|interface)([^{]*)[{]\\s*$')
```

Comment:
It pains me that these `Pattern`s are recreated on each invocation of this method. Please move them to static constants.